### PR TITLE
Issue #506

### LINKED ISSUE #506: Forbid multiline conditions
# Rule request

## Thesis

We need to forbid code like:

```python
       if isinstance(node.test, ast.UnaryOp) and isinstance(
            node.test.op,
            ast.Not,
        ):
```

It is really hard to read it. And conditions start to be "blurry".
The preferred way to write such conditions is:

```python
if isinstance(node.test, ast.UnaryOp:
    if isinstance(node.test.op, ast.Not):
```

## Reasoning

This way of writing conditions hides the inner complexity this line has.
And it decreases readability of the code.

### PR DESCRIPTION
**Pull request by shok9894**
Forbid multiline conditions, closes #506

### CODE REVIEW

File: tests/test_visitors/test_ast/test_conditions/test_multiline_conditions.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+incorrect_and_conditions1 = """if some and (
+    other == 1
+):
+    ...
+"""
+
+incorrect_or_conditions1 = """
+if some or (
+    other==1
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_conditions/test_multiline_conditions.py]: `other == 1`

File: tests/test_visitors/test_ast/test_conditions/test_multiline_conditions.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+):
+    ...
+"""
+
+incorrect_or_conditions2 = """
+if some or some_func(
+    other,
+):
+        ...
+"""
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_conditions/test_multiline_conditions.py]: Please, also include:
1. Correct conditions with multiple `and` and `or`
2. Correct conditions with `()` like function calls, prioritisation, and tuples
3. Incorrect conditions for multiple lines: 2, 3 and 4

File: wemake_python_styleguide/violations/best_practices.py [hunk 3]
```diff
@@ -1,1801 +1,1838 @@
...
-            except SomeException:
-                ...
-
-    .. versionadded:: 0.8.0
-
-    """
-
-    error_template = 'Found nested `try` block'
-    code = 464
+# -*- coding: utf-8 -*-
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/best_practices.py]: This diff is incorrect, please fix it.

File: tests/test_checker/test_noqa.py [hunk 4]
```diff
@@ -1,219 +1,220 @@
...
-        stdout=subprocess.PIPE,
-        stderr=subprocess.PIPE,
-        universal_newlines=True,
-        encoding='utf8',
-    )
-    stdout, _ = process.communicate()
-
-    for violation in IGNORED_VIOLATIONS:
-        assert stdout.count(violation) > 0
+# -*- coding: utf-8 -*-
```
Comment (sobolevn) [inline tests/test_checker/test_noqa.py]: This diff is incorrect, please fix it.

File: tests/fixtures/noqa.py [hunk 5]
```diff
@@ -1,380 +1,385 @@
...
-    print('Incorrect')
-
-try:
-    try:  # noqa: Z464
-        anti_z444 = 1
-    except ValueError:
-        raise TypeError('Second')
-except TypeError:
-    print('WTF?')
+# -*- coding: utf-8 -*-
```
Comment (sobolevn) [inline tests/fixtures/noqa.py]: This diff is incorrect, please fix it.

File: wemake_python_styleguide/violations/best_practices.py [hunk 6]
```diff
@@ -1,1801 +1,1838 @@
...
+    code = 431
+
+
+@final
+class MagicNumberViolation(ASTViolation):
+    """
+    Forbids to use magic numbers in your code.
+
+    What we call a "magic number"? Well, it is actually any number that
+    appears in your code ou
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/best_practices.py]: Also: use variables.

Comment (sobolevn) [inline wemake_python_styleguide/violations/best_practices.py]: 0.9.0

File: wemake_python_styleguide/visitors/ast/conditions.py [hunk 7]
```diff
@@ -1,59 +1,72 @@
...
-
-        Raises:
-            RedundantReturningElseViolation
-            NegatedConditionsViolation
-
-        """
-        self._check_negated_conditions(node)
-        self._check_redundant_else(node)
-        self.generic_visit(node)
+# -*- coding: utf-8 -*-
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/conditions.py]: This diff is incorrect, please fix it.

File: wemake_python_styleguide/visitors/ast/conditions.py [hunk 8]
```diff
@@ -1,59 +1,72 @@
...
+        if isinstance(node.test, ast.UnaryOp):
+            if isinstance(node.test.op, ast.Not):
+                self.add_violation(NegatedConditionsViolation(node))
+        elif isinstance(node.test, ast.Compare):
+            if any(isinstance(elem, ast.NotEq) for elem in node.test.ops):
+                self.add_violation(NegatedConditionsViolation(node))
+
+    def _check_multiline_conditions(self, node: ast.If) -> None:
+        """Checks multiline conditions ``if`` statement nodes."""
+        if isinstance(node.test, ast.BoolOp):
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/conditions.py]: In this case you are ignoring cases like:
```python
if very_long_call_name(
    long_parameter_name=long_variable_name,
):
```

Please, make sure it is covered and tested.

File: tests/test_visitors/test_ast/test_conditions/test_multiline_conditions.py (hunk 1 above)
Comment (shok9894) [reply tests/test_visitors/test_ast/test_conditions/test_multiline_conditions.py]: Complete

File: tests/test_visitors/test_ast/test_conditions/test_multiline_conditions.py (hunk 2 above)
Comment (shok9894) [reply tests/test_visitors/test_ast/test_conditions/test_multiline_conditions.py]: Complete

File: wemake_python_styleguide/violations/best_practices.py (hunk 3 above)
Comment (shok9894) [reply wemake_python_styleguide/violations/best_practices.py]: Fixed

File: tests/test_checker/test_noqa.py (hunk 4 above)
Comment (shok9894) [reply tests/test_checker/test_noqa.py]: Fixed

File: tests/fixtures/noqa.py (hunk 5 above)
Comment (shok9894) [reply tests/fixtures/noqa.py]: fixed

File: wemake_python_styleguide/violations/best_practices.py (hunk 6 above)
Comment (shok9894) [reply wemake_python_styleguide/violations/best_practices.py]: Complete

File: wemake_python_styleguide/visitors/ast/conditions.py (hunk 7 above)
Comment (shok9894) [reply wemake_python_styleguide/visitors/ast/conditions.py]: fixed

File: wemake_python_styleguide/visitors/ast/conditions.py (hunk 8 above)
Comment (shok9894) [reply wemake_python_styleguide/visitors/ast/conditions.py]: Change and tested it

File: tests/fixtures/noqa.py [hunk 9]
```diff
@@ -378,3 +378,40 @@ def bad_default_values(  # noqa: Z459
         raise TypeError('Second')
 except TypeError:
     print('WTF?')
+
+if some and (  # noqa: Z465
```
Comment (sobolevn) [inline tests/fixtures/noqa.py]: There's no need in putting all of these cases inside `noqa.py`.
It only contains integration tests. One is enough.
Move other cases into unit tests.

File: tests/test_checker/test_noqa.py [hunk 10]
```diff
@@ -151,6 +151,7 @@ def test_noqa_fixture_disabled(absolute_path, all_violations):
         'Z462': 1,
         'Z463': 1,
         'Z464': 1,
+        'Z465': 5,
```
Comment (sobolevn) [inline tests/test_checker/test_noqa.py]: 1

File: tests/test_visitors/test_ast/test_conditions/test_multiline_conditions.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    ...
+"""
+
+correct_conditions2 = """
+if some_func(k) and (some or other):
+    ...
+"""
+
+correct_conditions3 = """
+if (some_func(k) and some) or other in (1,2,3):
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_conditions/test_multiline_conditions.py]: Please, add:
```python
correct_conditions4 = """
if one:
    if two:
         ...
"""
```

**Pull request merged**